### PR TITLE
feat: support php 8.4 and other improvements

### LINKED ISSUE #246: feature: add support for PHP 8.4
The following PHP 8.4 features will require changes to the grammar:

- [x] https://github.com/php/php-src/pull/13029 [RFC](https://wiki.php.net/rfc/new_without_parentheses) 
- [x] https://github.com/php/php-src/pull/13455 [RFC](https://wiki.php.net/rfc/property-hooks)

### PR DESCRIPTION
**Pull request by calebdw**
Closes #246 

Changes:
- [PHP8.4] add support for `new` without parentheses
- [PHP8.4] add support for property hooks
- create `anonymous_class` rule
- rename `_variable_name` =>  `_simple_variable` to match php parser
- rename `_class_type_designator` => `_class_name_reference` to match php parser
- remove unnecessary semicolon from class declaration
- simplify `by_ref` definition
- create new `_identifier` and `_name` rules
- create `_dereferencable_scalar` rule to match php parser
- [BC] cleanup namespace rules
  - removed `namespace_name_as_prefix`, uses `prefix` field
  - removed `namespace_aliasing_clause`, uses `alias` field
  - removed `namespace_use_group_clause`, now uses `namespace_use_clause` inside `namespace_use_group`
- [BC] rename `anonymous_function_creation_expression` => `anonymous_function`
- remove inline rules to decrease state count (2627 => 2507)
- [BC] remove `property_initializer` rule, now uses a `default_value` field

Due to the BCs I plan to bump version to `0.23.0` after merging

### CODE REVIEW

**Comment by amaanq** [comment]
nice, I'll take a look soon

**Comment by aryx** [comment]
BTW @amaanq did you see my email? I tried to contact you on gmail but no answer :(

**Review by amaanq — APPROVED** [review]

**Comment by amaanq** [comment]
Bumping to 0.23.0 is fine, I only ask that you do that when I release 0.23.0 upstream (soon)

**Comment by amaanq** [comment]
thanks a ton again @calebdw, and I'm really sorry for taking a while :grimacing:

**Comment by calebdw** [comment]
No worries! Thanks for looking at it!

File: test/highlight/keywords.php [hunk 1]
```diff
@@ -14,7 +14,39 @@
...
+//^^^^^^^ @keyword
+//        ^ @module
+//          ^ @module
+
+use Foo\Baz as Baaz;
+//^ @keyword
+//  ^^^ @module
+//      ^^^ @type
+//          ^^ @keyword
+//             ^^^^ @type
```
Comment (verhovsky) [inline test/highlight/keywords.php:26]: but it's not a `type`.

**Pull request merged**